docs/architecture-overview.rst

```diff
@@ -0,0 +1,200 @@
+Zulip API clients.
+
+A server can host multiple Zulip *realms* (organizations) at the same
+domain, each of which is a private chamber with its own users,
+streams, customizations, and so on. This means that one person might
+be a user of multiple Zulip realms. The administrators of a realm can
+choose whether to allow anyone to register an account and join, or
+only allow people who have been invited, or restrict registrations to
+members of particular groups (using email domain names or corporate
+single-sign-on login for verification).
+
+
+Components
+==========
+
+
+Tornado and Django
+------------------
+
+Django is the main web application server; Tornado runs the
+server-to-client real-time push system. The app servers are configured
+by the Supervisor configuration (which explains how to start the
+server processes) and the nginx configuration (which explains which
+HTTP requests get sent to which app server).
+
+Tornado is an asynchronous server, so it can scale really well, but
+it's a real pain to write code for it. So Tornado gets only the parts
+that really need to scale because they involve maintaining a
+persistent connection from every single user. The parts that are
+activate relatively rarely (e.g. when people are typing) go into
+Django.
+
+
+nginx
+-----
+
+nginx is the front-end to all Zulip activity; it serves static assets
+and proxies to Django and Tornado. It handles HTTP requests according
+to the rules laid down in the many config files found in
+``zulip/puppet/zulip/files/nginx/``.
+
+``zulip/puppet/zulip/files/nginx/zulip-include-frontend/app`` is the
+most important of these files. It explains what happens when requests
+come in from outside.
+
+- All requests to URLs beginning with ``/static/`` are served from the
+  corresponding files in ``/home/zulip/prod-static/``.
+
+- Requests to ``/json/get_events``, ``/api/v1/events``, and
+  ``/sockjs`` are sent to the Tornado server running at
+  ``http://tornado``. These are requests to the real-time push system,
+  because the user's web browser sets up a long-lived TCP connection
+  with Tornado to serve as `a channel for push notifications <
+  https://en.wikipedia.org/wiki/Push_technology#Long_Polling>`__.
+
+- Requests to ``/api`` and to ``/`` (i.e., everything else) are sent
+  to the Django app via the UNIX socket
+  ``unix:/home/zulip/deployments/fastcgi-socket`` (defined in
+  ``puppet/zulip/files/nginx/zulip-include-frontend/upstreams``).
+
+
+
+Supervisor
+----------
+
+Starts processes and restarts them if they crash. The config file is
+``zulip/puppet/zulip/files/supervisor/conf.d/zulip.conf``. This is where
+Tornado and Django are set up, as well as a number of background
+processes that process event queues.
+
+
+memcached
+---------
+
+memcached is used to cache database model
+objects. ``zerver/lib/cache.py`` and ``zerver/lib/cache_helpers.py``
+manage putting things into memcached, and its configuration is in in
+``puppet/zulip/files/memcached.conf``.
```

Comment:
"managing putting things into memcached and invalidating the cache when values change"